runtime/src/main/java/org/corfudb/protocols/logprotocol/SMRRecord.java

```diff
@@ -43,81 +41,105 @@
-     *
+    /**
+     * An undo record, which can be used to undo this method.
      */
     @Getter
     public transient Object undoRecord;
 
-    /** A flag indicating whether an undo record is present. Necessary
+    /**
+     * A flag indicating whether an undo record is present. Necessary
      * because undo records may be NULL.
      */
     @Getter
     public boolean undoable;
 
-    /** The upcall result, if present. */
+    /**
+     * The upcall result, if present.
+     */
     @Getter
     public transient Object upcallResult;
 
-    /** If there is an upcall result for this modification. */
+    /**
+     * If there is an upcall result for this modification.
+     */
     @Getter
     public transient boolean haveUpcallResult = false;
 
-    /** Set the upcall result for this entry. */
+    /**
+     * The global address this record is associated to.
+     */
+    @Getter
+    @Setter
+    long globalAddress = Address.NON_ADDRESS;
+
+    /**
+     * Set the upcall result for this entry.
+     */
     public void setUpcallResult(Object result) {
         upcallResult = result;
         haveUpcallResult = true;
     }
 
-    /** Set the undo record for this entry. */
+    /**
+     * Set the undo record for this entry.
+     */
     public void setUndoRecord(Object object) {
         this.undoRecord = object;
         undoable = true;
     }
 
-    /** Clear the undo record for this entry. */
+    /**
+     * Clear the undo record for this entry.
+     */
     public void clearUndoRecord() {
         this.undoRecord = null;
         undoable = false;
     }
 
 
-    /** SMREntry constructor. */
-    public SMREntry(String smrMethod, @NonNull Object[] smrArguments, ISerializer serializer) {
-        super(LogEntryType.SMR);
+    /**
+     * SMRRecord constructor.
+     */
+    public SMRRecord(String smrMethod, @NonNull Object[] smrArguments, ISerializer serializer) {
         this.SMRMethod = smrMethod;
         this.SMRArguments = smrArguments;
         this.serializerType = serializer;
     }
 
     /**
-     * This function provides the remaining buffer. Child entries
-     * should initialize their contents based on the buffer.
+     * Deserialize from buffer and return a new {@code SMRRecord}.
      *
-     * @param b The remaining buffer.
+     * @param b byte buffer to deserialize.
+     * @return an {@code SMRRecord} deserialized from buffer.
      */
-    @Override
-    void deserializeBuffer(ByteBuf b, CorfuRuntime rt) {
-        super.deserializeBuffer(b, rt);
+    static SMRRecord deserializeFromBuffer(ByteBuf b, CorfuRuntime rt) {
+        SMRRecord record = new SMRRecord();
         short methodLength = b.readShort();
         byte[] methodBytes = new byte[methodLength];
         b.readBytes(methodBytes, 0, methodLength);
-        SMRMethod = new String(methodBytes);
-        serializerType = Serializers.getSerializer(b.readByte());
+        record.SMRMethod = new String(methodBytes);
```

Comment:
Explicitly set the Encoding ?

Reply:
Took another look at this after our discussion, since we are using netty's ByteBuf, I think netty takes care of charset?

Reply:
I don't think so, are you referring to endianness? that's different.

Reply:
Oh, I misunderstood. Done.